src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt

```diff
@@ -162,7 +158,14 @@ abstract class ManagedStrategy(
     // TODO: unify with `callStackTrace`
     // TODO: handle coroutine resumptions (i.e., unify with `suspendedFunctionsStack`)
     // TODO: extract into separate class
-    private var shadowStack = mutableThreadMapOf<ArrayList<ShadowStackFrame>>()
+    private val shadowStack = mutableThreadMapOf<ArrayList<ShadowStackFrame>>()
+
+    /**
+     * For each thread, stores a stack of entered method guarantees sections.
+     */
+    // TODO: unify with `shadowStack`
+    // TODO: handle coroutine resumptions (i.e., unify with `suspendedFunctionsStack`)
+    private val methodGuaranteesStack = mutableThreadMapOf<MutableList<ManagedGuaranteeType?>>()
```

Comment:
How does maintaining this additional data structure affect performance?

Reply:
Locally, when comparing with `develop`, there is no difference.